tensorflow_model_optimization/python/core/common/keras/compression/algorithms/bias_only_test.py

```diff
@@ -0,0 +1,217 @@
+  tflite_model = converter.convert()
+
+  with open(tflite_file, 'wb') as f:
+    f.write(tflite_model)
+
+  return tflite_file
+
+
+# TODO(tfmot): reuse test_utils_mnist.py.
+def _test_tflite(tflite_file):
+  interpreter = tf.lite.Interpreter(model_path=tflite_file)
+  interpreter.allocate_tensors()
+
+  input_index = interpreter.get_input_details()[0]['index']
+  output_index = interpreter.get_output_details()[0]['index']
+
+  (_, _), (x_test, y_test) = _get_dataset()
+
+  # Testing the entire dataset is too slow. Verifying only 300 of 10k samples.
+  x_test = x_test[0:300, :]
+  y_test = y_test[0:300]
+
+  total_seen = 0
+  num_correct = 0
+
+  for img, label in zip(x_test, y_test):
+    batch_input_shape = (1, 28, 28)
+    inp = img.reshape(batch_input_shape)
+    inp = inp.astype(np.float32)
+    total_seen += 1
+    interpreter.set_tensor(input_index, inp)
+    interpreter.invoke()
+    predictions = interpreter.get_tensor(output_index)
+
+    if np.argmax(predictions) == label:
+      num_correct += 1
+
+  return float(num_correct) / float(total_seen)
+
+
+def _get_directory_size_in_bytes(directory):
+  total = 0
+  try:
+    for entry in os.scandir(directory):
+      if entry.is_file():
+        # if it's a file, use stat() function
+        total += entry.stat().st_size
+      elif entry.is_dir():
+        # if it's a directory, recursively call this function
+        total += _get_directory_size_in_bytes(entry.path)
+  except NotADirectoryError:
+    # if `directory` isn't a directory, get the file size then
+    return os.path.getsize(directory)
+  except PermissionError:
+    # if for whatever reason we can't open the folder, return 0
+    return 0
+  return total
+
+
+class FunctionalTest(tf.test.TestCase):
+
+  def testBiasOnly_ReducesParamaters(self):
+    model = _build_model()
+    compressed_model = bias_only.optimize(model)
+
+    self.assertEqual(model.count_params(), 431080)
+    self.assertEqual(compressed_model.count_params(), 430508)
+
+  def testBiasOnly_HasReasonableAccuracy_TF(self):
+    model = _build_model()
+
+    compressed_model = bias_only.optimize(model)
+
+    _train_model(compressed_model)
+
+    _, (x_test, y_test) = _get_dataset()
+
+    loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+
+    compressed_model.compile(
+        optimizer='adam', loss=loss_fn, metrics=['accuracy'])
+
+    results = compressed_model.evaluate(x_test, y_test)
+
+    self.assertGreater(results[1], 0.60)
+
+  def testBiasOnly_HasReasonableAccuracy_TFLite(self):
+    model = _build_model()
+
+    compressed_model = bias_only.optimize(model)
+
+    _train_model(compressed_model)
+
+    saved_model_dir = _save_as_saved_model(compressed_model)
+    compressed_tflite_file = _convert_to_tflite(saved_model_dir)
+
+    accuracy = _test_tflite(compressed_tflite_file)
+
+    self.assertGreater(accuracy, 0.60)
+
+  # TODO(tfmot): can simplify to single layer test.
+  def testBiasOnly_BreaksDownLayerWeights(self):
```

Comment:
I think that the only test here that really adds test coverage is the one on preserving pretrained weights (which ends up going to _map_to_training_weights and doing something a bit different there).

I'd instead have this as a test under same_training_and_inference_test.py that just tweaks things from the default from same_training_and_inference.py so that the 'bias' is optimized instead. This tests something that the model developer would more realistically do (instead of creating a new algorithm) and reduces the amount of test code we need to have.



Everything else

Reply:
I think there's no way to compress 1-D tensor (bias) with SVD algorithm in my understanding. what's your expected behavior for that case?

Reply:
Ah I see. Yes there isn't a way to do so. 

I think my first comment on only the PreTrainedWeights adding coverage still holds, but we won't be able to use svd as I said.

Reply:
I see. I added comment that this algorithm can be removed when we find better algorithm to show bias compression or better coverage. (e.g. custom layer, custom weight compression)